pkg/containermeta/store.go

```diff
@@ -0,0 +1,332 @@
+	filter *Filter
+}
+
+// Store contains metadata for container workloads.
+type Store struct {
+	mu          sync.RWMutex
+	store       map[Kind]map[string]Entity
+	subscribers []subscriber
+
+	candidates map[string]Collector
+	collectors map[string]Collector
+
+	eventCh chan []Event
+}
+
+// NewStore creates a new container metadata store. Call Run to start it.
+func NewStore() *Store {
+	candidates := make(map[string]Collector)
+	for id, c := range collectorCatalog {
+		candidates[id] = c()
+	}
+
+	return &Store{
+		store:       make(map[Kind]map[string]Entity),
+		subscribers: []subscriber{},
+
+		candidates: candidates,
+		collectors: make(map[string]Collector),
+		eventCh:    make(chan []Event),
+	}
+}
+
+// Run starts the container metadata store.
+func (s *Store) Run(ctx context.Context) {
+	retryTicker := time.NewTicker(retryCollectorInterval)
+	pullTicker := time.NewTicker(pullCollectorInterval)
+	health := health.RegisterLiveness("containermeta-store")
+
+	// Dummy ctx and cancel func until the first pull starts
+	pullCtx, pullCancel := context.WithCancel(ctx)
+
+	log.Info("containermeta store initialized successfully")
+
+	go func() {
+		for {
+			select {
+			case <-health.C:
+
+			case <-pullTicker.C:
+				// pullCtx will always be expired at this point
+				// if pullTicker has the same duration as
+				// pullCtx, so we cancel just as good practice
+				pullCancel()
+
+				pullCtx, pullCancel = context.WithTimeout(ctx, pullCollectorInterval)
+				s.pull(pullCtx)
+
+			case evs := <-s.eventCh:
+				s.handleEvents(evs)
+
+			case <-retryTicker.C:
+				s.startCandidates(ctx)
+
+				if len(s.candidates) == 0 {
+					retryTicker.Stop()
+				}
+
+			case <-ctx.Done():
+				retryTicker.Stop()
+				pullTicker.Stop()
+
+				err := health.Deregister()
+				if err != nil {
+					log.Warnf("error de-registering health check: %s", err)
+				}
+
+				return
+			}
+		}
+	}()
+}
+
+// Subscribe returns a channel where container metadata events will be streamed
+// as they happen.
+func (s *Store) Subscribe(name string, filter *Filter) chan EventBundle {
+	// use an unbuffered channel since it's ok to block until the receiving
+	// end is free, and we need to wait for them to close the EventBundle's
+	// channel anyway.
+	ch := make(chan EventBundle)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.subscribers = append(s.subscribers, subscriber{
+		name:   name,
+		ch:     ch,
+		filter: filter,
+	})
+
+	if len(s.store) > 0 {
+		evs := []Event{}
+
+		for kind, entitiesOfKind := range s.store {
+			if !filter.MatchKind(kind) {
+				continue
+			}
+
+			// TODO(juliogreff): implement filtering by source
+
+			for _, entity := range entitiesOfKind {
+				evs = append(evs, Event{
+					Type:   EventTypeSet,
+					Entity: entity,
+				})
+			}
+		}
+
+		notifyChannel(ch, EventBundle{
+			Events: evs,
+			Ch:     make(chan struct{}),
+		})
+	}
+
+	return ch
+}
+
+func notifyChannel(ch chan EventBundle, bundle EventBundle) {
+	ch <- bundle
+	<-bundle.Ch
```

Comment:
I fear about the future developer experience, a not well-designed subscriber could basically deadlock or slow down all subscribers.

I know that we control the subscribers, but bugs can definitely happen. Note that we still hold the `mu` lock at this point, so we'll pile up events as well and subscriber cannot call `Get` or `List`.

Reply:
On reading from the store, it deadlocked as soon as I ran it, so I just pushed a fix. On badly implemented subscribers, we can either add a timeout and warn and move on when it is exceeded. It may also be an option to expose the blocking mechanism only to subscribers that want it, but I believe we can think about that once we have more subscribers.